app/assets/javascripts/discourse/tests/acceptance/user-test.js

```diff
@@ -131,3 +136,46 @@ acceptance("User Routes - Moderator viewing warnings", function (needs) {
     assert.ok($("div.alert-info").length, "has the permissions alert");
   });
 });
+
+acceptance("User - Saving user options", function (needs) {
+  needs.user({
+    admin: false,
+    moderator: false,
+    username: "eviltrout",
+    id: 1,
+    user_option: EmberObject.create({}),
+  });
+
+  needs.settings({
+    disable_mailing_list_mode: false,
+  });
+
+  test("saving user options", async function (assert) {
+    pretender.put("/u/eviltrout.json", () => {
+      return [
+        200,
+        { "Content-Type": "application/json" },
+        { success: true, user: {} },
+      ];
+    });
+    let spy = sinon.spy(User.current(), "_saveUserData");
+
+    await visit("/u/eviltrout/preferences/emails");
+    await click(".pref-mailing-list-mode input[type='checkbox']");
+    await click(".save-changes");
+
+    assert.ok(
+      spy.calledWithMatch({ mailing_list_mode: true }),
```

Comment:
Just leaving some breadcrumbs. If we had a choice we probably wouldn't use spy but I do not know a better choice here.

Reply:
@tgxworld I understand being against other mocks, but why not spy? The original function is still called, we just need to just know what it's called with. It's tricky because in this case the bug is in internal implementation code that doesn't return anything. The only other thing we can really do is moving the filtering of the user option fields into another function that returns the filtered fields, but not sure if that's worth disrupting the flow of the code for.

Reply:
Also we use spies all the time in ruby `expects(X).with(Y, Z)`

Reply:
In this case, the spy is spying on an internal private function so kind of similar to testing implementation instead of expected outcome. I did think about using pretender to filter for the request body and then setting a variable to indicate the request was called but I felt like it would make the code very complex.

Reply:
Yeah and I'm not sure if there is any way to observe/test what pretender actually sends. That would be really nice